### DISCUSSION TITLE
Implementation of moose surrogate models in multi-scale material problems

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#26161] [Q&A General]
Hi,
The problem that i was considering earlier (#19997) has now been stragetised to be implemented under a surrogate scheme in order to better manage the computational performance of the application we are developing for analyzing material problems in the crystal plasticity scale. The training data for the surrogates is planned to be derived from DAMASK homogenization modules, that comprises of the material deformation gradient and the corresponding homogenized stresses. The moose surrogate model then reads the macro deformation gradients from the component model and returns the stresses to their material quadrature points. The plan looks quite feasible give the features available in moose, but may need some tweaks to get in to correct form needed for the implementing the entire workflow. I would need the following to be answered to be able to start working on this proposed model set-up.
1) I am not quite clear on the format of the training data that i should be writing out from my damask full model that contains deformation gradients and stresses. Looking at the existing code ('NearestPointSurrogate.C'), i see that we might need some features for reading this training data file (*.rd), as currently they can only take in-built data coming from the sampler objects. I welcome any suggestions here, not to forget the format for the training dataset.
2) The other important thing to be answered here is the communication between the component model QP's and the surrogate model. The surrogates are really quick and so it would be more compelling to go for a moose multiapp way of implementation, where the existing 'QuadraturePointMultiapp' [(#21340)](https://mooseframework.inl.gov/source/multiapps/QuadraturePointMultiApp.html) can be exploited here to handle the surrogate reference from and to the component model. Not quite sure if this is the correct way to address this coupling, but welcome better ideas/suggestions.

The workflow can be visualized here for your reference.
![image](https://github.com/idaholab/moose/assets/40111479/140d35d2-a314-4d2d-8bec-c94c53034fed)

Kind regards,
Arun

**Comment by GiudGiud** [comment]
Hello

What is this `.rd` format? Is this standardized in some way and we could have a reader for it in MOOSE?
Could it be converted to a simple csv file?
@grmnptr on whether we could support a file format here instead
Guillaume

**Reply by manikumd22** [reply]
Hi. @GiudGiud ,

I am working on the same multiscale problem on MOOSE in my case I have CSV file that will take F and will give P. But I'm struggling with the implementation in MOOSE.
Can you please hep me with this?

**Reply by GiudGiud** [reply]
Hello

There's plenty of utilities for parsing a CSV file. For example the DelimitedFileReader utility.
Then once the CSV file is in memory you can use the reader to query interpolated data, or use other utilities (MultiDimensionalInterpolation notably) to interpolate within the CSV data

**Reply by manikumd22** [reply]
Hi @GiudGiud


I am new to MOOSE, I am struggling with the same kind of problem like reading the CSV file and then using the values in the material. I am reading strain CSV file and I am using it to calculate _S[_qp] every gauss point. Is it correct way to write it. Please let me know

I write it the way you described

InputParameters
ComputeNeoHookeanStress::validParams()
{
  InputParameters params = ComputeLagrangianStressPK2::validParams();

  params.addParam<MaterialPropertyName>("lambda",
                                        "lambda",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("mu",
                                        "mu",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");

  params.addParam<MaterialPropertyName>("strain_csv","This strain will be used to calculate the stress tensor");

  return params;
}

ComputeNeoHookeanStress::ComputeNeoHookeanStress(const InputParameters & parameters)
  : ComputeLagrangianStressPK2(parameters),
    _lambda(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda"))),
    _mu(getMaterialProperty<Real>(getParam<MaterialPropertyName>("mu")))
    _strain_csv(getMaterialProperty<Real>(getParam<MaterialPropertyName>("strain_csv")))

{
}

void
ComputeNeoHookeanStress::computeQpPK2Stress()
{
  // Hyperelasticity is weird, we need to branch on the type of update if we
  // want a truly linear model
  //
  // This is because we need to drop quadratic terms for the linear update
  usingTensorIndices(i_, j_, k_, l_);

  // Large deformation = nonlinear strain
  if (_large_kinematics)
  {
    RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
    _S[_qp] = (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
    _C[_qp] =
        -2 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
        _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv);
  }
  // Small deformations = linear strain
  else
  {
    const auto I = RankTwoTensor::Identity();
    RankTwoTensor strain = 0.5 * (_F[_qp] + _F[_qp].transpose()) - I;
    _C[_qp] = _lambda[_qp] * I.times<i_, j_, k_, l_>(I) +
              2.0 * _mu[_qp] * RankFourTensor(RankFourTensor::initIdentitySymmetricFour);
    strain_PBC = _strain_csv
    _S[_qp] = strain_PBC;
  }
}

**Comment by grmnptr** [comment]
Surrogates that are trained within moose are saved in an `rd` file. We support reading csv tables and creating surrogates from those as well. The surrogate itself, however will be saved in a restartable data file so that we can load it easily. 

There is one surrogate which goes beyond this. We have an interface to libtorch so those could potentially be loaded in a torchscript format.

Edit: we have a surrogate system with well-defined interfaces. So you can fetch/evaluate a surrogate super easily.

There have been multiple works which used surrogates in material models. I would suggest looking at the `evaluate` function.
